src/deploy/functions/runtimes/python/index.ts

```diff
@@ -0,0 +1,199 @@
+import fetch from "node-fetch";
+import * as path from "path";
+import * as fs from "fs";
+import * as portfinder from "portfinder";
+import * as spawn from "cross-spawn";
+import * as cp from "child_process";
+
+import * as runtimes from "..";
+import * as backend from "../../backend";
+import { logger } from "../../../../logger";
+import * as discovery from "../discovery";
+import { FirebaseError } from "../../../../error";
+import { Build } from "../../build";
+
+export const LATEST_VERSION: runtimes.Runtime = "python39";
+
+export const PYVENV = "venv";
+
+const AUTOGENERATED_FOLDER = "firebase";
+const AUTOGENERATED_FILE = "admin";
+
+type ChildProcessWithCompletionPromise = cp.ChildProcess & { promise: Promise<string> };
+
+/**
+ * Runs a shell command with the Python virtual env pre-activated.
+ */
+export function runWithVirtualEnv(
+  commandAndArgs: string[],
+  functionsDir: string,
+  withVirtualEnv = true,
+  opts?: cp.SpawnOptions
+): ChildProcessWithCompletionPromise {
+  const activateScriptPath =
+    process.platform === "win32" ? ["Scripts", "activate.bat"] : ["bin", "activate"];
+  const venvActivate = path.join(functionsDir, PYVENV, ...activateScriptPath);
+  const command = withVirtualEnv
+    ? process.platform === "win32"
+      ? venvActivate
+      : "source"
+    : commandAndArgs[0];
+  const args = withVirtualEnv
+    ? [process.platform === "win32" ? "" : venvActivate, "&&", ...commandAndArgs]
+    : [...commandAndArgs.splice(1)];
+
+  const child = spawn(command, args, {
+    shell: true,
+    cwd: functionsDir,
+    stdio: [/* stdin= */ "ignore", /* stdout= */ "pipe", /* stderr= */ "inherit"],
+    ...opts,
+    env: {
+      ...(opts?.env ?? {}),
+      ...process.env,
+    },
+  }) as ChildProcessWithCompletionPromise;
+  let out = "";
+  child.stdout?.on("data", (chunk: Buffer) => {
+    const chunkString = chunk.toString();
+    out = out + chunkString;
+    logger.debug(chunkString);
+  });
+  const promise = new Promise<string>((resolve, reject) => {
+    child.on("exit", () => resolve(out));
+    child.on("error", reject);
+  });
+  child.promise = promise;
+  return child;
+}
+
+class Delegate implements runtimes.RuntimeDelegate {
+  public readonly name = "python";
+  constructor(
+    private readonly projectId: string,
+    private readonly sourceDir: string,
+    public readonly runtime: runtimes.Runtime
+  ) {}
+
+  private modulesDir_ = "";
+
+  async modulesDir(): Promise<string> {
+    if (!this.modulesDir_) {
+      const out = await runWithVirtualEnv(
+        [
+          "python3.9",
```

Comment:
Above `LATEST_VERSION` is defined as `python39`, but here it's `python3.9` - just confirming that's intended.

Reply:
Correct, "python39" is the GCP runtime name, "python3.9" is the actual python binary name